src/libsyntax/parse/parser/diagnostics.rs

```diff
@@ -916,6 +916,54 @@ impl<'a> Parser<'a> {
         Err(err)
     }
 
+    pub(super) fn expect_semi(&mut self) -> PResult<'a, ()> {
+        if self.eat(&token::Semi) {
+            return Ok(());
+        }
+        let sm = self.sess.source_map();
+        if self.look_ahead(1, |t| {
+            (t == &token::CloseDelim(token::Brace) || token_can_begin_expr(t)) && t.kind != token::Colon
+        }) && [token::Comma, token::Colon].contains(&self.token.kind) {
+            // Likely typo: `,` → `;`, `:` → `;`.
+            match (sm.lookup_line(self.prev_span.lo()), sm.lookup_line(self.token.span.lo())) {
+                (Ok(ref a), Ok(ref b)) if a.line != b.line => {
```

Comment:
Elaborate on all the various conditions that lead here would be good & include an example that illustrates why this heuristic is used.